src/libopensc/card-dnie.c

```diff
@@ -729,6 +729,62 @@ int dnie_match_card(struct sc_card *card)
 	LOG_FUNC_RETURN(card->ctx, result);
 }
 
+static int dnie_sm_free_wrapped_apdu(struct sc_card *card,
+		struct sc_apdu *plain, struct sc_apdu **sm_apdu)
+{
+	struct sc_context *ctx = card->ctx;
+	int rv = SC_SUCCESS;
+
+	LOG_FUNC_CALLED(ctx);
+	if (!sm_apdu)
+		LOG_FUNC_RETURN(ctx, SC_ERROR_INVALID_ARGUMENTS);
+	if (!(*sm_apdu))
+		LOG_FUNC_RETURN(ctx, SC_SUCCESS);
+
+	plain->resp = (*sm_apdu)->resp;
+	plain->resplen = (*sm_apdu)->resplen;
+	plain->sw1 = (*sm_apdu)->sw1;
+	plain->sw2 = (*sm_apdu)->sw2;
```

Comment:
Thanks a lot for reviewing this so quickly!!!

DNIe does have authenticated status bytes, however they are not used here. I know this is a bit twisted, but it is the only way I have managed to make it work so far. dnie_sm_free_wrapped_apdu doesn't actually decrypt the APDU, it just copies the encrypted response and the status bytes of the encrypted APDU to "plain" (which I have just noticed, it is an awful choice of a name). Decryption still takes place in the end of dnie_transmit_apdu, cwa-dnie.c:825.
My final goal here is to remove dnie_transmit_apdu entirely and put all wrapping and unwrapping in dnie_sm_free_wrapped_apdu and dnie_sm_get_wrapped_apdu in the card-dnie.c module. I use the same names for the static methods in cwa-dnie.c and card-dnie.c, for me it is not too bad since they perform the same task and the goal is to move code from those in cwa-dnie.c to those in card-dnie.c, but I should have clarified this.
As it is right now:
- dnie_sm_free_wrapped_apdu and dnie_sm_get_wrapped_apdu in card-dnie.c just copy the encrypted APDU back and forth. They are use from the OpenSC SM code.
- dnie_sm_get_wrapped_apdu in cwa-dnie.c encodes the APDU and modifies some of the fields. It is called from dnie_transmit_apdu.
- dnie_sm_free_wrapped_apdu in cwa-dnie.c copies the response to the plain APDU, which is decoded in dnie_transmit_apdu, although decoding could be moved within this method without problems. It is also called from dnie_transmit_apdu.
It is clearly not the prettiest pull request I have done, but it has taken me around a month to reach this far, and I would like to get some (working) partial result in before continuing. Your input on how to proceed is highly appreciated.